### ISSUE TITLE
Configuration cache issues

### ISSUE THREAD

**Issue by matejdro** [j-roskopf/ComposeGuard#40]
Hello,

since installing this plugin we are experiencing a configuration cache issue where removing the compose_reports folder will completely invalidate configuration cache:

`Calculating task graph as configuration cache cannot be reused because the file system entry 'my-module/build/compose_reports' has been removed.`
 
This folder shouldn't be included in configuration, it should only be part of task execution.

**Comment by matejdro** [comment]
I think I can reproduce it by doing this:



1. Run <variant>ComposeCompilerCheck gradle task (with configuration cache enabled)
2. Delete `build` folder of that module
3. Run <variant>ComposeCompilerCheck gradle task again

EDIT: No it does not seem to be that. Will keep trying.

**Comment by matejdro** [comment]
I have managed to find a reliable reproducer:

1. Delete `.gradle/configuration-cache` folder to start from a fresh slate
2. Run `compileDebugKotlin --configuration-cache`
3. Run `debugComposeCompilerCheck --configuration-cache`
4. Delete `my-module/build/compose_reports`
2. Run `compileDebugKotlin --configuration-cache`
3. Run `debugComposeCompilerCheck --configuration-cache`

BUG: Build will not reuse configuration cache and will have to re-configure with warning described above.

**Comment by j-roskopf** [comment]
Hey! Sorry for the delayed response. For some reason I didn't get any notification to my email. Thank you very much for the detailed reproducer! Will take a look shortly.

**Comment by j-roskopf** [comment]
thanks again for the report. i got a solution going locally, should have a PR up shortly

**Comment by matejdro** [comment]
I cannot reproduce this one anymore, thanks!

**Comment by matejdro** [comment]
I managed to reproduce it again, albeit with a slightly different error message:

1. Delete `.gradle/configuration-cache` folder to start from a fresh slate
2. Run `debugComposeCompilerCheck --configuration-cache`
3. Delete `my-module/build`
4. Run `debugComposeCompilerCheck --configuration-cache` again

BUG: Build will not reuse configuration cache with `Calculating task graph as configuration cache cannot be reused because an input to plugin 'com.android.internal.library' has changed.` warning.

**Comment by j-roskopf** [comment]
I'm not entirely sure the above issue is related to this plugin.

If I repro the steps but just deleting the `my-module/build/compose_reports` folder, it works as expected

`com.android.internal.library` isn't anything that is familiar to me. I see this youtrack link for a similar issue related to a bug in kapt that gives the same message https://youtrack.jetbrains.com/issue/KT-58167/Applying-KAPT-plugin-to-Gradle-8.1-causes-configuration-cache-miss

I was able to repro the steps above in my sample project, but when I removed ksp (and the one library depending on ksp) from that module, I was no longer able to repro. I added back just ksp, and I was not able to repro. It was only after adding back the one library depending on ksp did I see the behavior again

This is the library I added to my sample project to help verify ksp related stuff, so I'm thinking more it's either ksp / some other dependency related?

https://github.com/bennyhuo/KotlinDeepCopy

**Comment by matejdro** [comment]
I think this might be caused by using `beforeVariants`: https://github.com/j-roskopf/ComposeGuard/blob/51ec02458598cc52bffae9d3c2e585ccf57cac17/compose-guard/gradle-plugin/src/main/kotlin/com/joetr/compose/guard/ReportGenPlugin.kt#L391

Probably Gradle considers everything used inside this block as a dependency to the Android plugin?

**Comment by j-roskopf** [comment]
I'm not entirely sure, that could be true, but that still wouldn't explain why when removing KSP and ksp dependencies from my test project (while still applying compose guard) the issue goes away. can you repro the issue with no annotation processing and compose guard applied?

**Comment by matejdro** [comment]
You are right, I did some debugging and it does appear to be KSP-related:

![image](https://github.com/user-attachments/assets/5af971a1-524d-4546-a74f-a17877ac1625)

It is weird that it happens only with compose guard, but I guess we can't do much here. Thanks for looking into it.

**Comment by matejdro** [comment]
https://github.com/j-roskopf/ComposeGuard/blob/d966181e0471cd6548b77aa6cd83204a1879bef5/compose-guard/gradle-plugin/src/main/kotlin/com/joetr/compose/guard/ReportGenPlugin.kt#L217

I think this line might cause some more issues, because it checks whether directory is not empty in configuration time

**Comment by matejdro** [comment]
Yup, I can reproduce:

1. Delete `my-module/compose_reports`
2. Delete `.gradle/configuration-cache` folder to start from a fresh slate
3. Run `compileDebugKotlin --configuration-cache`
4. Create `my-module/compose_reports` folder
5. Run `compileDebugKotlin --configuration-cache` again

Configuration cache fails with `Calculating task graph as configuration cache cannot be reused because directory 'my-module/compose_reports' has changed.` 

This is with Compose Guard `2.3.4`

**Comment by matejdro** [comment]
I think these lines may also be problematic:

https://github.com/j-roskopf/ComposeGuard/blob/d966181e0471cd6548b77aa6cd83204a1879bef5/compose-guard/gradle-plugin/src/main/kotlin/com/joetr/compose/guard/ReportGenPlugin.kt#L149

https://github.com/j-roskopf/ComposeGuard/blob/d966181e0471cd6548b77aa6cd83204a1879bef5/compose-guard/gradle-plugin/src/main/kotlin/com/joetr/compose/guard/ReportGenPlugin.kt#L183

Basically, you should not access any file/folder properties during configuration time, you can only do this during execution time.

**Comment by j-roskopf** [comment]
am I doing something incorrectly in the steps to repro? I am unfortunately unable to repro in my sandbox app 

https://github.com/user-attachments/assets/d3870e79-bb82-43dd-821b-06f91699302f

I am not really sure how to get around reading the files at configuration time to have a fully functioning plugin. Let's take the first referenced example

```
                            val isGenDirectoryNotEmpty = genExtension.outputDirectory.get().list()?.isNotEmpty() == true
                            t.outputs.upToDateWhen {
                                isGenDirectoryNotEmpty
                            }
```

This code was added to work around the scenario where you generate the metrics, delete the folder, and then generate the metrics again. Without the above code, the metrics would not get re-regenerated because no kotlin src code has changed, so the kotlin compile task stays up to date. 

I could change it so the the gen output directory is an input to the kotlin compile task via the code below, but that would seem to still be doing the same thing, unless the offending part is truly the `.list()` on the directory

```
            tasks.withType(KotlinJvmCompile::class.java).configureEach(
                object : Action<KotlinJvmCompile> {
                    override fun execute(t: KotlinJvmCompile) {
                        val isCheckTask =
                            gradle.startParameter.taskNames.any {
                                it.contains(CHECK_TASK_NAME)
                            }
                        val isWriteTask =
                            gradle.startParameter.taskNames.any {
                                it.contains(GENERATE_TASK_NAME)
                            }
                        if (composeGuardExtension.configureKotlinTasks.get() && (isWriteTask || isCheckTask)) {
                            t.inputs.files(genExtension.outputDirectory.get())
                        }
                    }
                },
            )
```

**Comment by matejdro** [comment]
Above does not reproduce anymore with `2.3.5`, because you added checks for write or check task. But it still reproduces if you use `debugComposeCompilerCheck` instead of `compileDebugKotlin`.

Above genDirectoryNotEmpty example is luckily very easy to fix. You just have to move the whole check to the `upToDateWhen` block:

```
                            t.outputs.upToDateWhen {
                                val isGenDirectoryNotEmpty = genExtension.outputDirectory.get().list()?.isNotEmpty() == true
                                isGenDirectoryNotEmpty
                            }
```

`upToDateWhen` block is executed at execution time, not configuration time, so it does not violate the configuration cache anymore.

**Comment by j-roskopf** [comment]
will push out the change with 2.3.6

```
                            t.outputs.upToDateWhen {
                                val isGenDirectoryNotEmpty = genExtension.outputDirectory.get().list()?.isNotEmpty() == true
                                isGenDirectoryNotEmpty
                            }
```

**Comment by matejdro** [comment]
I can still reproduce with `2.3.6`, presumably because of the two leftover folder accesses that I mentioned above.

Repro project: [ConfigCacheDemo.zip](https://github.com/user-attachments/files/17009195/ConfigCacheDemo.zip) (essentially just a default Android hello world template, with Compose Guard added in).

1. Delete `app/compose_reports`
2. Delete `.gradle/configuration-cache` folder to start from a fresh slate
3. Run `./gradlew debugComposeCompilerCheck --configuration-cache`
4. Create `app/compose_reports` folder
5. Run `./gradlew debugComposeCompilerCheck --configuration-cache` again

Configuration cache fails with `Calculating task graph as configuration cache cannot be reused because directory 'app/compose_reports' has changed.`

**Comment by j-roskopf** [comment]
The repro project wasn't in a good state to repro.

running `debugComposeCompilerCheck` without configuring the check task to report all on missing baseline

```
composeGuardCheck {
    reportAllOnMissingBaseline.set(true)
}
```

would result in an error of no baseline file. After adding that configuration, there are dynamic properties that exist in the sample project

```
* What went wrong:
Execution failed for task ':app:debugComposeCompilerCheck'.
> New @dynamic parameters were added! 
  FunctionAndParameter(functionName=ConfigCacheDemoTheme, parameterName=darkTheme, parameterType=Boolean)
  More info: https://github.com/JetBrains/kotlin/blob/master/plugins/compose/design/compiler-metrics.md#default-parameter-expressions-that-are-dynamic
```

I modified the theme to remove support for dark mode and now see the `Calculating task graph as configuration cache cannot be reused because the file system entry 'app/compose_reports' has been created.`

I am no able to repro with 2.3.7, can you give it a try please?

**Comment by matejdro** [comment]
Result of the task does not matter. For me, I could reproduce the configuration cache warning on `2.3.6` regardless of the missing baseline error, so I did not bother fixing it.

But `2.3.7` seems to resolve the issue, thanks!
